lib/Sema/TypeCheckType.cpp

```diff
@@ -1260,31 +1329,6 @@ resolveTopLevelIdentTypeComponent(TypeResolution resolution,
     if (options.contains(TypeResolutionFlags::SilenceErrors))
       return ErrorType::get(ctx);
 
-    if (comp->getIdentifier() == ctx.Id_Self &&
```

Comment:
One way of doing this would be to keep this code in here, but then check the method signature after the fact for invalid occurrences of Self. This is better in the long run since the current code for Self returns in classes is hacky and I was hoping your PR could eventually replace it entirely.

Reply:
I'm not qualified for any major overhaul. I'm just trying to get `Self` working for purely Self-ish reasons (ducks)

Reply:
One last thing worth mentioning is one leg of the patch uses `getDeclaredInterfaceType()` and the other `getSelfInterfaceType()` should they use the same?

Reply:
You should prefer getSelfInterfaceType(). They differ only in the case of a protocol though, and your code won't be reached with protocols because protocols have a generic parameter named Self that is found by ordinary name lookup.

Reply:
Thanks, Switched to getSelfInterfaceType() for both branches.